### DISCUSSION TITLE
Integration in a loopback4 microservice

### DISCUSSION THREAD

**Discussion by ingpconci** [bpmnServer/bpmn-server#66] [Ideas]
I try to create an extension for the loopback microservice framework using a Postgresql Database as the backend.
In the first experiment, I copied the source code from bpmn-server (ver 1.1.10) and modified some files and filenames.
I created the loopback model mapped with Postgresl
```
@model({
  settings: {idInjection: false, postgresql: {schema: 'appuser', table: 'bpmnProcessInstance'}}
})
export class BpmnProcessInstance extends Entity {
  @property({
    type: 'string',
    id: true,
    postgresql: {columnName: 'id', dataType: 'character varying', dataLength: 128, dataPrecision: null, dataScale: null, nullable: 'NO'},
  })
  id?: string;
  @property({
    type: 'string',
    required: true,
    length: 255,
    jsonSchema: {nullable: false},
    postgresql: {columnName: 'name', dataType: 'character varying', dataLength: 255, dataPrecision: null, dataScale: null, nullable: 'NO'},
  })
  name: string;
........

```
I exposed some methods using the loopback4 endpoint.
Example:
 
```
  //---------------------------------------------------------------------
  // Invoke the execution of a task
  //---------------------------------------------------------------------
  @authorize({
    allowedRoles: [RoleKeys.ADMINISTRATOR]
  })
  @post('/bpmn-process-instances/execute-task')
  @response(200, {
    description: 'BpmnProcessInstance execute task result',
    content: {
      'application/json': {
        schema: {
          type: 'object',
          properties: {
            requestResultMessage: {
              type: 'string',
            },
          },
        },
      },
    },
  })
  async executeProcessInstanceTask(
    @requestBody({
      content: {
        'application/json': {
          schema: getModelSchemaRef(ExecuteTaskInput, {
            title: 'ExecuteTaskInput Data'
          }),
        },
      },
    })
    theExecuteTaskInput: ExecuteTaskInput,
  ): Promise<{requestResultMessage: string}> {
    const processInstanceItemId = theExecuteTaskInput.processInstanceItemId;
    const inputData = theExecuteTaskInput.dataJson;
    const resultInvokeItem = await this.bpmnEngineService.invokeProcessInstanceItem(processInstanceItemId,inputData);
    console.log('BpmnProcessInstanceController:POST /bpmn-process-instances/execute-task:  resultInvokeItem=', resultInvokeItem);
   .......

```
To support parallel-gateway I modified gateway.ts.
To facilitate the debug I added the debug library

To integrate in loopback4 services using the main bpmn-server repository, I need to :
- rename files: example DataStore.ts TO data-store.ts
- support custom datastore (example: 
```
var records = await this.db.find(this.dbConfiguration.db, Instance_collection, query, projection);
substitute from code like this
bpmnProcessInstanceArray = await this.server.bpmnProcessInstanceRepository.find();
```
It is feasible to modify the code of bpmn-server to facilitate the integration in the loopback server framework?
I want to share my loopback4-bpmnserver-component in GitHub and contribute to bpmn-server project.

Regards
Paolo

**Comment by ralphhanna** [comment]
Hi Paolo,
I am not familiar with loopback4 so I you need to help me out here.
Not sure why do you have to use  Postgresql instead of current implementation of MongoDB?

But to answer your question about supporting custom datastore:
1. Modify configuration.ts
```js
		dataStore: function (server) {
			return new CustomDataStore(server);
		},
```
2. Create own CustomDataStore:
```js
class CustomDataStore extends ServerComponent  implements IDataStore {
}
```
3. Implement your own logic 

Also, you are more than welcome to contribute to bpmn-server through pull-request.

Thanks,
Ralph

**Reply by ingpconci** [reply]
Thanks for the suggestion.
I've finished the modification to use Postgresql as DB.
I've modified these files.
**interfaces/common.ts**
Added configuration for DB loopback4 DB connection 
```
        {
            db_url: string,
            db: string,
        }
        },
        loopbackRepositories?:any
    },
```
**common\DefaultConfiguration.ts**

```
    database: { MongoDB: { db_url: string; db: string; }; };
    database: { 
        MongoDB: { db_url: string; db: string; }; 
        loopbackRepositories?:any;
    };
    logger: ILogger;
    apiKey: string;
```
**server\Engine.ts**
There are parts of the code that do not permit to use CustomDatasource
Row 41
```
        // new dataStore for every execution to be monitored 
        const newDataStore =new DataStore(this.server);
        // const newDataStore =new DataStore(this.server);
        const newDataStore = this.server.configuration.dataStore(this.server);
        this.server.dataStore = newDataStore;
        newDataStore.monitorExecution(execution);
```
Row 95

```
            // new dataStore for every execution to be monitored 
            const newDataStore = new DataStore(execution.server);
            //const newDataStore = new DataStore(execution.server);
            const newDataStore = this.server.configuration.dataStore(execution.server);
            execution.server.dataStore = newDataStore;
```
In my loopback4 server, I implemented the Custom datasources and register in the configuration file.
**configurations.ts**

```
		database: {
			MongoDB:
			{
				db_url: "mongodb://localhost:27017?retryWrites=true&w=majority",  //db_url: process.env.MONGO_DB_URL,
				db: 'bpmn',//'bpmn'  process.env.MONGO_DB_NAME
			},
			loopbackRepositories: {
			}
		},
		apiKey: '', //process.env.API_KEY
		/* Define Server Services */
		logger: function (server: any) {
			new Logger(server);
		},
		definitions: function (server: BPMNServer) {
			return new CustomModelsDatastore(server);
		},
		appDelegate: function (server: any) {
			return new MyAppDelegate(server);
		},
		dataStore: function (server: any) {
			return new CustomDataStore(server);
		},
```

In the startup part of my server, I create the instance of BPMNServer.

```
    // attach repositories instances
    configuration.database.loopbackRepositories.bpmnProcessInstanceRepository = this.bpmnProcessInstanceRepository;
    configuration.database.loopbackRepositories.bpmnProcessModelRepository = this.bpmnProcessModelRepository;
    // create the server
    this.bpmnServer = new BPMNServer(configuration, new Logger({toConsole: false, toFile: '', callback: null}));
```

Now I found a bug on parallelGateway, but before modifying the code, it is better to work after a pull request.
![image](https://user-images.githubusercontent.com/8488830/194568007-166b5b94-164e-4af5-9339-fbef3093ca37.png)

I have a question about the code organization; why are the typescript compiled js code in the same dir of source ts code?

Paolo

**Reply by ralphhanna** [reply]
1. Code that reference DataSource directly, is no longer needed , remove in release 1.3.11
2. Bug with parallelGateway , please send me the bpmn file or import it to bpmn.omniworkflow.com
3.  Please request a Pull-Request for your changes and will review and incorporate
4. Location of compile js: what do you recommend?

Thanks

**Reply by ingpconci** [reply]
I imported the example to bpmn.omniworkflow.com.
The model name is test_gateway1.bpmn
![image](https://user-images.githubusercontent.com/8488830/194648142-12ed4deb-9d71-451f-af98-586e988ae519.png)


I try to make a Pull-Request; 
![image](https://user-images.githubusercontent.com/8488830/194647415-39fef6a7-fa66-487e-9a44-fe931dd69953.png)

But I don't know if I need to create a Fork to make the request or if I need an open Issue.

Regarding the location of compiles js , I suggest a dist folder
![image](https://user-images.githubusercontent.com/8488830/194648594-d2d95a1b-2204-4c6b-af5a-3e33cac8595b.png)

**Comment by ralphhanna** [comment]
To create a pull request make a fork

Sent from my iPhone

On Oct 7, 2022, at 4:37 PM, ingpconci ***@***.***> wrote:

﻿

I imported the example to bpmn.omniworkflow.com.
The model name is test_gateway1.bpmn
[image]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F8488830%2F194648142-12ed4deb-9d71-451f-af98-586e988ae519.png&data=05%7C01%7C%7C4766a31e6ec44409438208daa8a3d018%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638007718765819564%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=yvWglCsQeLsHXmKg%2FzcNC0v6gDK%2FqCA6M90iqzWlpeg%3D&reserved=0>

I try to make a Pull-Request;
[image]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F8488830%2F194647415-39fef6a7-fa66-487e-9a44-fe931dd69953.png&data=05%7C01%7C%7C4766a31e6ec44409438208daa8a3d018%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638007718766132044%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=CJjP0kyED6aiKepaWfZgPofK5YCmZ0PjQP%2FhCr98lxE%3D&reserved=0>

But I don't know if I need to create a Fork to make the request or if I need an open Issue.

Regarding the location of compiles js , I suggest a dist folder
[image]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F8488830%2F194648594-d2d95a1b-2204-4c6b-af5a-3e33cac8595b.png&data=05%7C01%7C%7C4766a31e6ec44409438208daa8a3d018%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638007718766132044%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=RR236GgjCuJVsobDgX9DI1djSmw5xzjKWfnFt6u00jE%3D&reserved=0>

**Comment by ingpconci** [comment]
I finished creating a pull request
